### ISSUE TITLE
The future of Wrapper Versioning, updates, and Release Pipeline

### ISSUE THREAD

**Issue by JasonBristol** [benbaptist/minecraft-wrapper#299]
Fully understanding the potential pushback from this one, I would still like to mark this issue clearly so as to have a logical and in depth discussion with the core focus on the health and quality of Wrapper.py.
(EDIT - see discussion background below)
Here is what I propose for Feature Pack 1:

- [x] Adopt Semantic Versioning as our versioning standard
- [x] Refactor our versioning implementation to leverage the full breadth of version.py
- [ ] Refactor automatic updates to fetch directly through git via git remote update and git show-branch *<branch>
- [ ] Assume that checksum is desired for public external repositories and fetch from HEAD

These changes serve more than just cosmetic purposes, this enables us to do one of several things for the future:

- [x] Re-point reference repositories for automatic updates, this gives users the ability to pull updates from alternative sources such as forks [...].
- [x] Automatic versioning via commit, i.e. each commit would update the full version as enabled by version.py, this allows for fine grain versioning and could help build test driven plugins for automated testing.
- [ ] Automated changelogs

(Original discussion)
[discus..?]

While many may have differing opinions on what version scheme to use, we should not shy away from discussing a real versioning scheme that works toward quality automation within wrapper and external systems.

If you are not familiar with Semantic Versioning I suggest giving it a serious look as it is leveraged by a lion's share of modern codebases and has been adopted as the de-facto standard.

As of today Wrapper.py sits at version [0.8.1] -- this in semantic versioning terms means that it has not even hit an official stable release yet (I think we agree that this is accurate). With the intent [stabilizing] Wrapper.py, as well as incorporate features that some would grant as required to provide value in various areas, we should consider a v1.0.0-rc1 release.

Abiding by our branch philosophy would mean that we could develop our development branch to an indefinite point, providing a PR to BenBaptist/Development whenever necessary. Ben could very well log on to have a PR that advances Wrapper far beyond just simple bug fixes, but rather a backlog of changes that mature Wrapper. This is okay, source control implemented correctly mandates a division that does not risk touching the quasi "golden state" of BenBaptist/Master until explicitly desired.

With that in mind, and adopting semantic versioning as our version schema, I have already staged utils.version.py in our experimental branch for further development. PEP compliant and borrowed from Django, version.py will allow us to automatically rev, fetch, and compare versions in a way that utterly deprecates facilities like build_script.py.

Before the issue of the MD5 checksum comes up, lets review the use case of such a checksum.

When Wrapper.py automatically looks for and fetches updates, as of today, it sends a raw HTTP request to github to read out our build/ directory and globals.py. It leverages a combination of comparing the checksum as well as the version number to validate that an update is available. This does work, but let me suggest a more robust and modular way that is hinted at in version.py and is leveraged by many other mature implementations:

The very act of using git as a source control engine means that each and every commit has a checksum generated. Why store our own poor man's checksum when we can simply leverage this functionality through say git log --pretty=format:%ct --quiet -1 HEAD?

Further to this point, is a checksum even needed for automatic updates at all? If we were to remote fetch any development or master updates through git, we would not need to worry about the integrity of the content.

**Comment by suresttexas00** [comment]
Sorry, I accidentally wrote this back by comitting directly: https://github.com/suresttexas00/minecraft-wrapper/commit/149d67ed0c039ca78607b2c758fb0380a4d23ed8


(first time using git to push a commit- until now I did everything on github)

**Comment by suresttexas00** [comment]
Revisiting this again.  About : `git log --pretty=format:%ct --quiet -1 HEAD`... This is not a checksum, but a timestamp.  Do you know how to get the actual checksum?

**Comment by suresttexas00** [comment]
To answer my own question (and end up raising another), so either of these work:

```
git log -1 --format="%H"
git log --pretty=format:'%h' -n 1
```
but what is the point of getting this checksum if we can't independently verify it.

Frankly, I think this is all pointless in this day and time since an attacker could  (if we were hosting our own makeshift repo or something- which we are not) simply modify the Wrapper.py and our hash to match.  With no signature schema, it seems pointless.

**Comment by suresttexas00** [comment]
A bit more thought on this is that the checksum is less about security and more about ensuring the Wrapper.py file did not get corrupted in the requests download process.. and therefore continues to be a desirable thing.  If we ever implement (bind ourselves to) github such that we depend on the git engine to deliver source changes (and move to running the source as our primary mode), then a checksum may be superfluous.

**Comment by suresttexas00** [comment]
I did not wan't to but I did just release a development 1.0.0 "beta", which will be converted to a 1.0.0 "rc" once the more serious remaining bugs are worked out.  If we like it, I will move that to the stable/master as 1.0.0. 'final'.  I pushed the version because I had finished integrating the version.py into wrapper's build_script.py and core.wrapper update schema.  I was going to wait to push those changes en-mass, along with the other fixes, but Ben has showed back up and I need to keep my repo consistent with this one to prevent possible confusion if Ben starts producing pull requests.
